### PR TITLE
Copter: improved follow at close distances

### PR DESCRIPTION
**Pull request by tridge**
we need to add the target velocity after we scale for being close to the target or we end up with a large offset lag
This makes a huge difference with 3 copters following with offsets of 5m

### CODE REVIEW

File: ArduCopter/mode_follow.cpp [hunk 1]
```diff
@@ -104,13 +93,24 @@ void ModeFollow::run()
...
-        desired_velocity_neu_cms.y = desired_velocity_xy_cms.y;
+        desired_velocity_neu_cms.xy() = desired_velocity_xy_cms;
 
         // limit vertical desired_velocity_neu_cms to slow as we approach target (we use 1/2 of maximum deceleration for gentle slow down)
         const float des_vel_z_max = copter.avoid.get_max_speed(pos_control->get_pos_z_p().kP().get(), pos_control->get_max_accel_z_cmss() * 0.5f, fabsf(dist_vec_offs_neu.z), copter.G_Dt);
         desired_velocity_neu_cms.z = constrain_float(desired_velocity_neu_cms.z, -des_vel_z_max, des_vel_z_max);
 
+        /*
+          add velocity of the target. We add this at the end as we do
+          not want to scale it when we get close to the target
```
Comment (Georacer) [inline ArduCopter/mode_follow.cpp:104]: I'd say "We add this at the end as we don't want to apply the slowdown calculations to the baseline target velocity too" is a more accurate reasoning. How close we are to the target doesn't play a role in the slowdown calculations.

File: ArduCopter/mode_follow.cpp [hunk 2]
```diff
@@ -68,18 +68,7 @@ void ModeFollow::run()
...
-        if (!is_zero(desired_speed_xy) && (desired_speed_xy > pos_control->get_max_speed_xy_cms())) {
-            const float scalar_xy = pos_control->get_max_speed_xy_cms() / desired_speed_xy;
-            desired_velocity_neu_cms.x *= scalar_xy;
-            desired_velocity_neu_cms.y *= scalar_xy;
-            desired_speed_xy = pos_control->get_max_speed_xy_cms();
-        }
+        desired_velocity_neu_cms = dist_vec_offs_neu * kp;
 
         // limit desired velocity to be between maximum climb and descent rates
         desired_velocity_neu_cms.z = constrain_float(desired_velocity_neu_cms.z, -fabsf(pos_control->get_max_speed_down_cms()), pos_control->get_max_speed_up_cms());
```
Comment (Georacer) [inline ArduCopter/mode_follow.cpp:74]: This vertical velocity limitation shouldn't be here, before the slowdown calculations, but after ~L112, like the XY limitations.

**Review by Georacer — CHANGES_REQUESTED** [review]
I believe this is not correct yet, as the Z-velocity limitation to `get_max_speed_down_cms()` hasn't been moved after the slowdown calculations and the addition of the target velocity (like the XY component have).

Other than that, I proposed a small comment re-wording.

Finally, not strictly part of the code changes, but I think the part at L76-81 is unused, so it'd be nice to delete it:
```cpp
        // unit vector towards target position (i.e. vector to lead vehicle + offset)
        Vector3f dir_to_target_neu = dist_vec_offs_neu;
        const float dir_to_target_neu_len = dir_to_target_neu.length();
        if (!is_zero(dir_to_target_neu_len)) {
            dir_to_target_neu /= dir_to_target_neu_len;
        }
```

**Comment by tridge** [comment]
closing in deference to https://github.com/ArduPilot/ardupilot/pull/26142

**Pull request closed without merging**